larray/core/session.py

```diff
@@ -1475,6 +1478,173 @@ def display(k, v, is_metadata=False):
         return res
 
 
+class ArrayDef(ABCArray):
+    def __init__(self, axes=None):
+        if axes is not None and not isinstance(axes, AxisCollection):
+            axes = AxisCollection(axes)
+        self.axes = axes
+
+
+class ConstrainedSession(Session):
+    """
+    Examples
+    --------
+
+    >>> class ModelVariables(ConstrainedSession):
+    ...     # declare constant (read-only) variables. Their values cannot be changed at runtime.
+    ...     FIRST_OBS_YEAR = 1991
+    ...     LAST_PROJ_YEAR = 2070
+    ...     AGE = Axis('age=0..120')
+    ...     GENDER = Axis('gender=male,female')
+    ...     G_CHILDREN = AGE[:17]
+    ...     G_ADULTS = AGE[18:]
+    ...     # declare variables with defined types.
+    ...     # Their values will be defined at runtime but must match the specified type.
+    ...     variant_name = str
+    ...     first_proj_year = int
+    ...     obs_years = Axis
+    ...     proj_years = Axis
+    ...     population_obs = Array
+    ...     population_proj = Array
+    ...     # declare arrays with fixed axes
+    ...     mortality_rate = ArrayDef((AGE, GENDER))
```

Comment:
Also, I think it that it would be useful to provide a way for users to declare that an array must have some specific axes even if the labels of those axes are not specified yet. In the example above, this would mean the ability to write something like either:
```python
population_obs = ArrayDef((AGE, GENDER, "year"))
```
or better yet (if possible -- I think it would be tricky but possible):
```python
population_obs = ArrayDef((AGE, GENDER, obs_years))
```
the later might require using AxisDef **instances** instead of just "Axis" to be able to differentiate instances: eg:
```python
obs_years = AxisDef("time")
population_obs = ArrayDef((AGE, GENDER, obs_years))
```

The goal is keep the class definition as close as possible to a complete reference of what each variable is (and also be more precise in the safety checks)

Reply:
I also wonder how PyCharm behaves when using "show definition" on such an attribute. Does it show the class definition or the point where it was given a value "m.obs_years = Axis(...)"? I am guessing the later, when it is defined but I don't really know.

Reply:
> Also, I think it that it would be useful to provide a way for users to declare that an array must have some specific axes even if the labels of those axes are not specified yet.

You're right but can I move this suggestion to a separate issue and implement it later? This PR is complicated enough.